_posts/2022-07-24-opensearch-plugin-zips-now-in-maven-repo.markdown

```diff
@@ -0,0 +1,74 @@
+date: 2022-07-24 01:01:01 -0700
+categories: 
+  - technical-post
+twittercard:
+  description: "Details on how to consume OpenSearch plugin zips from maven repo and the process involved in shipping them to the maven repo."
+---
+
+### Motivation and Current Challenges:
+
+The plugins zips that are used as dependencies cannot be downloaded dynamically during runtime, as they are not part of version controlled maven system. Only the respective Java jars are available to download using maven coordinates. This forces a user to now to use a dependency plugin zip of some random version built on a developer desktop, which is not reliable, also the tests executed as part of this build zip cannot be accurate, so eventually for few plugins the local build zip is forced to check in to the plugin repo to facilitate the plugin availability for dependency, this is one of the challenge faced by the community. Coming to the download part of plugins as separate isolated components via a cached mechanism like maven is not possible, with this challenge users had to  pre-bake the random built zip to the project.
+
+Starting with release `2.1.0` OpenSearch plugin zips are now signed and published to central maven repo, with `groupID` as [org.opensearch.plugin](https://repo1.maven.org/maven2/org/opensearch/plugin/). These zips can now be fetched as individual components by directly downloading using maven coordinates or by clickstream from central maven repo which can be later cached to the desired user local maven repo. The snapshot version of the zips can be fetched from nexus maven repo with the same `groupID` as [org.opensearch.plugin](https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/plugin/).
+
+### Maven publish task with custom gradle plugin.
+
+The publishing of zips to central maven repo had been possible with custom gradle plugin `opensearch.pluginzip`.
+Starting with release `2.1.0` all OpenSearch gradle supported plugins will have custom gradle plugin `opensearch.pluginzip` applied to the gradle project, this will create a new task `publishPluginZipPublicationToZipStagingRepository` which does all the heavy lifting in identifying the distribution plugin zip, setting the maven coordinates, generating the POM file, updating the user passed POM fields and publishing it to maven repo. More details and inner workings of the plugin can found at [opensearch-plugins](https://github.com/opensearch-project/opensearch-plugins/blob/main/BUILDING.md#opensearchpluginzip).
+
+
+**Figure 1**: Workflow that ship generated plugin zips to maven
+
+![Figure 1: Workflow that ship zips to maven]({{ site.baseurl }}/assets/media/blog-images/2022-07-24-opensearch-plugin-zips-now-in-maven-repo/figure1.png){: .img-fluid }
+
+### Benefits of plugin zips in maven:
+
+* The plugin dependencies can now be fetched using the right maven coordinates from maven repo and a user need not use a zip of some random version built on a developer desktop.
```

Comment:
```suggestion
## Maven zip publication and gradle

OpenSearch publishes plugin zips using a custom gradle plugin `opensearch.pluginzip`. With OpenSearch 2.1, all OpenSearch gradle supported plugins create a new task, `publishPluginZipPublicationToZipStagingRepository`. The task performs all the heavy lifting for users by:

- Identifying the distribution plugin zip
- Setting the Maven coordinates
- Generating the POM file
- Updating the user-generated PM fields
- Publishing the task to your Maven repo

You can find more details about the inner workings of OpenSearch plugins in the [opensearch-plugins repo](https://github.com/opensearch-project/opensearch-plugins/blob/main/BUILDING.md#opensearchpluginzip).

**Figure**: Workflow that ship generated plugin zips to maven

![Figure: Workflow that ship zips to maven]({{ site.baseurl }}/assets/media/blog-images/2022-07-24-opensearch-plugin-zips-now-in-maven-repo/figure1.png){: .img-fluid }
```

Reply:
Heading change as `Maven zip publication with Gradle`

Add `2.1` in code quotes.

A small adjustment in 4th point as follows:
- Updating the user passed POM fields

Reply:
Capitalize all instances of "Gradle". Line 33: Add a comma after "OpenSearch publishes plugin zips using a custom Gradle plugin,". Change to "Gradle-supported". Add terminating periods to all list items. The fourth list item should read "Updating the user-generated POM fields." The Figure caption should read "Workflow that ships generated plugin zips to Maven".